hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -554,10 +613,300 @@ struct TypeVisitor : visitor::PreOrder<bool, TypeVisitor> {
                 // Nothing.
                 break;
         }
+
+        return false;
+    }
+
+    result_t operator()(const type::ValueReference& x, position_t p) {
+        switch ( _stage ) {
+            case Stage::COLLECT: {
+                const auto& type_id = x.typeID();
+
+                if ( ! type_id )
+                    break;
+
+                // Record this type as used.
+                _used[*type_id] = true;
+                break;
+            }
+            case Stage::PRUNE_USES:
+            case Stage::PRUNE_DECLS:
+                // Nothing.
+                break;
+        }
+
         return false;
     }
 };
 
+struct ConstantFoldingVisitor : OptimizerVisitor, visitor::PreOrder<bool, ConstantFoldingVisitor> {
+    // TODO(bbannier): Index constants by their canonical ID once it is
+    // available. We should also be able to remove `Node::rid` at that point.
+    std::map<uint64_t, bool> _constants;
+
+    void collect(Node& node) override {
+        _stage = Stage::COLLECT;
+
+        for ( auto i : this->walk(&node) )
+            dispatch(i);
+    }
+
+    bool prune_uses(Node& node) override {
+        _stage = Stage::PRUNE_USES;
+
+        bool any_modification = false;
+
+        while ( true ) {
+            bool modified = false;
+            for ( auto i : this->walk(&node) ) {
+                if ( auto x = dispatch(i) )
+                    modified = *x || modified;
+            }
+
+            if ( ! modified )
+                break;
+
+            any_modification = true;
+        }
+
+        return any_modification;
+    }
+
+    bool operator()(const declaration::GlobalVariable& x, position_t p) {
+        // We only work on boolean constants.
+        if ( ! (x.isConstant() || x.type() == type::Bool()) )
+            return false;
+
+        switch ( _stage ) {
+            case Stage::COLLECT: {
+                const auto& init = x.init();
+                assert(init);
+
+                if ( auto ctor = init.value().tryAs<expression::Ctor>() )
+                    if ( auto bool_ = ctor->ctor().tryAs<ctor::Bool>() )
+                        _constants[p.node.rid()] = bool_->value();
+
+                break;
+            }
+
+            case Stage::PRUNE_USES:
+            case Stage::PRUNE_DECLS: break;
+        }
+
+        return false;
+    }
+
+    bool operator()(const expression::ResolvedID& x, position_t p) {
+        switch ( _stage ) {
+            case Stage::COLLECT:
+            case Stage::PRUNE_DECLS: return false;
+            case Stage::PRUNE_USES: {
+                auto rid = x.rid();
+
+                if ( const auto& constant = _constants.find(rid); constant != _constants.end() ) {
+                    if ( x.type() == type::Bool() ) {
+                        HILTI_DEBUG(logging::debug::GlobalOptimizer, util::fmt("inlining constant '%s'", x.id()));
+
+                        replaceNode(p, builder::bool_(constant->second));
```

Comment:
If the `GlobalVariable` is under a `Module` after this replacement its old value will be cached in the `Module`'s `declarations`; we should clear that cache after this edit to make sure we see consistent results.